src/clients/c/samples/main.c

```diff
@@ -0,0 +1,435 @@
+        }
+    }
+
+    // Releasing the packet, so it can be used in a next request.
+    release_packet(&packets_pool, packet);
+
+    // Cleanup
+    completion_context_destroy(&ctx);
+    tb_client_deinit(client);
+}
+
+tb_packet_t* acquire_packet(tb_packet_list_t *packet_list) {
+    
+    // This sample is single-threaded,
+    // In real use, this function should be thread-safe.
+    tb_packet_t *packet = packet_list->head;
+
+    if (packet == NULL) {
+        printf("Too many concurrent requests\n");
+        exit(-1);
+    }
+
+    packet_list->head = packet->next;
+    packet->next = NULL;
+    
+    if (packet_list->head == NULL) {
+        packet_list->tail = NULL;
+    }
+
+    return packet;
+}
+
+void release_packet(tb_packet_list_t *packet_list, tb_packet_t *packet) {
+    // This sample is single-threaded,
+    // In real use, this function should be thread-safe.
+    if (packet_list->head == NULL) {
+        packet_list->head = packet;
+        packet_list->tail = packet;
+    } else {
+        packet_list->tail->next = packet;
+        packet_list->tail = packet;
+    }
+}
+
+#if IS_POSIX
+
+void on_completion(
+    uintptr_t context, 
+    tb_client_t client, 
+    tb_packet_t *packet, 
+    const uint8_t *data, 
+    uint32_t size
+) {    
+    // The user_data gives context to a request:
+    completion_context_t* ctx = (completion_context_t*)packet->user_data;
+
+    // Signaling the main thread we received the reply:
+    pthread_mutex_lock(&ctx->lock);
+
+    memcpy (ctx->reply, data, size);
+    ctx->size = size;
+    ctx->completed = true;
+    
+    pthread_cond_signal(&ctx->cv);
+    pthread_mutex_unlock(&ctx->lock);
+}
+
+void send_request(
+    tb_client_t client,
+    tb_packet_list_t *packets,
+    completion_context_t *ctx
+) {
+    // Locks the mutex:
+    if (pthread_mutex_lock(&ctx->lock) != 0) {
+        printf("Failed to lock mutex\n");
+        exit(-1);
+    }
+
+    // Submits the request asynchronously:
+    ctx->completed = false;
+    tb_client_submit(client, packets);
+
+    // Uses a condvar to sync this thread with the callback:
+    while (!ctx->completed) {
+        if (pthread_cond_wait(&ctx->cv, &ctx->lock) != 0) {
+            printf("Failed to wait condvar\n");
+            exit(-1);
+        }
+    }
+    
+    if (pthread_mutex_unlock(&ctx->lock) != 0) {
+        printf("Failed to unlock mutex\n");
+        exit(-1);
+    }
+}
+
+void completion_context_init(completion_context_t *ctx) {
+    if (pthread_mutex_init(&ctx->lock, NULL) != 0) {
+        printf("Failed to initialize mutex\n");
+        exit(-1);
+    }
+
+    if (pthread_cond_init(&ctx->cv, NULL) != 0) {
+        printf("Failed to initialize condition var\n");
+        exit(-1);
+    }
+}
+
+void completion_context_destroy(completion_context_t *ctx) {
+    pthread_cond_destroy(&ctx->cv);
+    pthread_mutex_destroy(&ctx->lock);
+}
+
+long long get_time_ms(void) {
+    struct timeval tv;
+    if (gettimeofday(&tv,NULL) != 0) {
```

Comment:
should probably use `clock_gettime(CLOCK_MONOTONIC)` instead. Unlike `gettimeofday`, it can't be affected by changes to system time either by the user, ntp-sync, or dst.